fullnode/fullnode_cache.go

```diff
@@ -0,0 +1,228 @@
+	return &cachingFullNode{
+		logger:             logger,
+		lazyFullNode:       lazyFullNode,
+		sessionCache:       sessionCache,
+		accountPubKeyCache: accountPubKeyCache,
+	}, nil
+}
+
+// GetApp is a NoOp (apps fetched only at startup; relaying fetches sessions for app/session sync).
+func (cfn *cachingFullNode) GetApp(ctx context.Context, appAddr string) (*apptypes.Application, error) {
+	return nil, fmt.Errorf("GetApp is a NoOp in the caching full node")
+}
+
+// GetSession returns (and auto-refreshes) the session for a service/app from cache.
+func (cfn *cachingFullNode) GetSession(
+	ctx context.Context,
+	serviceID sdk.ServiceID,
+	appAddr string,
+) (sessiontypes.Session, error) {
+	// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#get-or-fetch
+	return cfn.sessionCache.GetOrFetch(
+		ctx,
+		getSessionCacheKey(serviceID, appAddr),
+		func(fetchCtx context.Context) (sessiontypes.Session, error) {
+			cfn.logger.Debug().Str("session_key", getSessionCacheKey(serviceID, appAddr)).Msgf(
+				"[cachingFullNode.GetSession] Making request to full node",
+			)
+			return cfn.lazyFullNode.GetSession(fetchCtx, serviceID, appAddr)
+		},
+	)
+}
+
+// getSessionCacheKey builds a unique cache key for session: <prefix>:<serviceID>:<appAddr>
+func getSessionCacheKey(serviceID sdk.ServiceID, appAddr string) string {
+	return fmt.Sprintf("%s:%s:%s", sessionCacheKeyPrefix, serviceID, appAddr)
+}
+
+// GetAccountPubKey returns the account public key for the given address.
+// The cache has no TTL, so the public key is cached indefinitely.
+//
+// The `fetchFn` param of `GetOrFetch` is only called once per address on startup.
+func (cfn *cachingFullNode) GetAccountPubKey(
+	ctx context.Context,
+	address string,
+) (pubKey cryptotypes.PubKey, err error) {
+	// See: https://github.com/viccon/sturdyc?tab=readme-ov-file#get-or-fetch
+	return cfn.accountPubKeyCache.GetOrFetch(
+		ctx,
+		getAccountPubKeyCacheKey(address),
+		func(fetchCtx context.Context) (cryptotypes.PubKey, error) {
+			cfn.logger.Debug().Str("account_key", getAccountPubKeyCacheKey(address)).Msgf(
+				"[cachingFullNode.GetPubKeyFromAddress] Making request to full node",
+			)
+			return cfn.lazyFullNode.GetAccountPubKey(fetchCtx, address)
+		},
+	)
+}
+
+// getAccountPubKeyCacheKey returns the cache key for the given account address.
+// It uses the accountPubKeyCacheKeyPrefix and the account address to create a unique key.
+//
+// eg. "pubkey:pokt1up7zlytnmvlsuxzpzvlrta95347w322adsxslw"
+func getAccountPubKeyCacheKey(address string) string {
+	return fmt.Sprintf("%s:%s", accountPubKeyCacheKeyPrefix, address)
+}
+
+// ValidateRelayResponse: passthrough to underlying node.
+func (cfn *cachingFullNode) ValidateRelayResponse(
+	ctx context.Context,
+	supplierAddr sdk.SupplierAddress,
+	responseBz []byte,
+) (*servicetypes.RelayResponse, error) {
+	return cfn.lazyFullNode.ValidateRelayResponse(ctx, supplierAddr, responseBz)
+}
```

Comment:
Note how we're delegating functionaliy from one full node to another for something relay related.

Can you think you of a way to refactor this to `relay.go` file has a `ValidateRelayResponse` function?

Reply:
It's moved to the new GatewayClient struct in the new PR.